proteinworkshop/features/node_features.py

```diff
@@ -153,12 +153,17 @@ def compute_surface_feat(
 @jaxtyped
 @beartype
 def orientations(
-    X: Union[CoordTensor, AtomTensor], ca_idx: int = 1
+    X: Union[CoordTensor, AtomTensor], batch_index: torch.Tensor, ca_idx: int = 1
 ) -> OrientationTensor:
     if X.ndim == 3:
         X = X[:, ca_idx, :]
-    forward = _normalize(X[1:] - X[:-1])
-    backward = _normalize(X[:-1] - X[1:])
-    forward = F.pad(forward, [0, 0, 0, 1])
-    backward = F.pad(backward, [0, 0, 1, 0])
-    return torch.cat((forward.unsqueeze(-2), backward.unsqueeze(-2)), dim=-2)
+    batched_orientations = []
+    for batch_id in batch_index.unique(sorted=True):
```

Comment:
I think this can be vectorised by using indices from `batch._slice_dict["coords"] to set the relevant indices to 0

Reply:
OK. Have a look at this new (slightly longer) vectorized implementation: https://github.com/a-r-j/ProteinWorkshop/pull/58/commits/a1d864bb00d5538c99ae42bfcb5595b34e1a5e47

Reply:
Truthfully, all of the asserts (commented out for now) should probably be made into a unit test to detect issues like this in the future.